File: react-native-landmarks-id-lod/ios/LandmarksIdLOD.m
//
//  LandmarksIdLOD.m
//  LandmarksIdLod
//
//  Copyright © 2022 Facebook. All rights reserved.
//

#import <Foundation/Foundation.h>
#import <React/RCTBridgeModule.h>

@interface RCT_EXTERN_MODULE(RNLandmarksIdLOD, NSObject)

    RCT_EXTERN_METHOD(initialiseSDK:(NSString*)appId appSecret:(NSString*)appSecret apiKey:(NSString*)apiKey debugMode:(BOOL)debugMode);

    RCT_EXTERN_METHOD(startTracking);
    RCT_EXTERN_METHOD(stopTracking);
    RCT_EXTERN_METHOD(terminate);

    RCT_EXTERN_METHOD(requestLocationPermission);

    RCT_EXTERN_METHOD(restartRecordingData);
    RCT_EXTERN_METHOD(stopRecordingData);

    RCT_EXTERN_METHOD(setCustomerId:(NSString*)customerId);
    RCT_EXTERN_METHOD(sendCustomString:(NSString*)key value:(NSString *)value);
    RCT_EXTERN_METHOD(sendCustomInteger:(NSString*)key value:(nonnull NSNumber *)value);
    RCT_EXTERN_METHOD(sendCustomFloat:(NSString*)key value:(nonnull NSNumber *)value);

@end

File: react-native-landmarks-id-lod/ios/LandmarksIdLOD.swift
//
//  LandmarksIdLOD.swift
//  LandmarksIdLod
//
//  Copyright © 2022 Facebook. All rights reserved.
//

import Foundation
//import LandmarksIDSDK

@objc(RNLandmarksIdLOD)
class RNLandmarksIdLOD: NSObject {
    
    // MARK: - Properties
    
    var landmarksIdManager: LandmarksIDManagerDelegate? = LandmarksIDManagerDelegate.sharedManager
    
    // MARK: - Initialiser
    
    override init() {
        print("SDK class initialised");
    }
    
    // MARK: - Functions
    
    @objc static func requiresMainQueueSetup() -> Bool {
        return false
    }
    
    // sdk initialisers
    
    @objc func initialiseSDK(_ appId: String, appSecret: String, apiKey: String, debugMode: Bool) -> Void {
        DispatchQueue.main.async {
            LandmarksIDManagerDelegate.initialize(appId, appSecret: appSecret, apiKey: apiKey)
            self.landmarksIdManager?.debug = debugMode
            self.landmarksIdManager?.discoveryEnabled = true
            self.landmarksIdManager?.setup()
        }
    }
    
    // tracking
    
    @objc func startTracking() {
        landmarksIdManager?.startTracking()
    }
    
    @objc func stopTracking() {
        landmarksIdManager?.stopTracking()
    }
    
    @objc func terminate() {
        landmarksIdManager?.applicationWillTerminate()
    }
    
    // location
    
    @objc func requestLocationPermission() {
        landmarksIdManager?.requestLocationPermissions(.authorizedAlways)
    }
    
    // device level data collection
    
    @objc func restartRecordingData() {
        landmarksIdManager?.restartRecordingData()
    }
    
    @objc func stopRecordingData() {
        landmarksIdManager?.stopRecordingData()
    }
    
    // send additional data
    
    @objc func setCustomerId(_ customerId: String) {
        landmarksIdManager?.customerId = customerId
    }
    
    @objc func sendCustomString(_ key: String, value: String) {
        landmarksIdManager?.setCustom(stringValue: value, key: key)
    }
    
    @objc func sendCustomInteger(_ key: String, value: NSNumber) {
        landmarksIdManager?.setCustom(intValue: value, key: key)
    }
    
    @objc func sendCustomFloat(_ key: String, value: NSNumber) {
        landmarksIdManager?.setCustom(floatValue: value, key: key)
    }
}
